gpt_engineer/ai.py

```diff
@@ -23,72 +39,134 @@ class TokenUsage:
 
-    def start(self, system, user, step_name):
+    def start(self, system: str, user: str, step_name: str) -> List[Message]:
         messages = [
-            {"role": "system", "content": system},
-            {"role": "user", "content": user},
+            SystemMessage(content=system),
+            HumanMessage(content=user),
         ]
-
         return self.next(messages, step_name=step_name)
 
-    def fsystem(self, msg):
-        return {"role": "system", "content": msg}
+    def fsystem(self, msg: str) -> SystemMessage:
+        return SystemMessage(content=msg)
 
-    def fuser(self, msg):
-        return {"role": "user", "content": msg}
+    def fuser(self, msg: str) -> HumanMessage:
+        return HumanMessage(content=msg)
 
-    def fassistant(self, msg):
-        return {"role": "assistant", "content": msg}
+    def fassistant(self, msg: str) -> AIMessage:
+        return AIMessage(content=msg)
 
-    def next(self, messages: List[Dict[str, str]], prompt=None, *, step_name=None):
+    def combine_messages(self, messages: List[Message]) -> str:
+        msg_dict = messages_to_dict(messages)
+        logging.debug(msg_dict)
+        prompt = "\n".join(
+            "%s: %s" % (md["type"], md["data"]["content"]) for md in msg_dict
+        )
+        logging.debug("Prompt: " + prompt)
+        return prompt
+
+    def next(
+        self,
+        messages: List[Message],
+        prompt: Optional[str] = None,
+        *,
+        step_name: str,
+    ) -> List[Message]:
         if prompt:
-            messages += [{"role": "user", "content": prompt}]
+            messages.append(self.fuser(prompt))
 
         logger.debug(f"Creating a new chat completion: {messages}")
-        response = openai.ChatCompletion.create(
-            messages=messages,
-            stream=True,
-            model=self.model,
-            temperature=self.temperature,
-        )
 
-        chat = []
-        for chunk in response:
-            delta = chunk["choices"][0]["delta"]  # type: ignore
-            msg = delta.get("content", "")
-            print(msg, end="")
-            chat.append(msg)
-        print()
-        messages += [{"role": "assistant", "content": "".join(chat)}]
+        msgs_as_prompt = self.combine_messages(messages)
+        response = self.llm(msgs_as_prompt, callbacks=[StreamingStdOutCallbackHandler()])
+        response = cleanup_reponse(response)
+
+        messages.append(self.fassistant(response))
+
         logger.debug(f"Chat completion finished: {messages}")
 
         self.update_token_usage_log(
-            messages=messages, answer="".join(chat), step_name=step_name
+            messages=messages, answer=response, step_name=step_name
         )
 
         return messages
 
-    def update_token_usage_log(self, messages, answer, step_name):
+    def last_message_content(self, messages: List[Message]) -> str:
+        m = messages[-1].content.strip()
+        logging.info(m)
+        return m
```

Comment:
I think this function makes the code more difficult to read for a human.

Would prefer to remove it!

And `logging.info(m)` will be confusing no?

Reply:
1. Agree & deleted `logging.info(m)`
2. Disagree with `last_message_content` making the code more difficult to read. We're using that function 11 times (see image), so think makes sense to not always rewrite it, and the function name is pretty self-explaining IMO. But will do what you decide. Lmk!

![image](https://github.com/AntonOsika/gpt-engineer/assets/40663591/661dc0dd-e457-463d-9520-3a42cc271720)

Reply:
Great on 1.

On the other, it's just that it's:

1. More characters to type
2. Needs an external dependency (AI)
3. which is an object? (even though it doesn't need to be initialized)
4. hides away unnecessarily, how simple everything actually is: It's just a list of messages! (knowing this makes it easier to understand the code)

Reply:
Okay, get your point, will change!